File: apps/android/app/src/main/java/com/mdelacour/mynotes/ai/tools/Journal.kt
package com.mdelacour.mynotes.ai.tools

import kotlinx.serialization.Serializable

@Serializable
data class AgentEditRecord(
	val fromUtf16: Int,
	val toUtf16: Int,
	val deleted: String,
	val inserted: String,
	val startAnchor: String,
	val endAnchor: String,
)

@Serializable
data class MutationRecord(
	val id: String,
	val journalId: String,
	val exchangeId: String,
	val messageId: String,
	val toolCallId: String,
	val createdAt: Long,
	val kind: String,
	val noteId: String,
	val beforeRevision: String? = null,
	val afterRevision: String? = null,
	val revertState: String = "pending",
	val revertNote: String? = null,
	val edits: List<AgentEditRecord> = emptyList(),
	val createdContent: String? = null,
	val deletedContent: String? = null,
	val deletedOrderIndex: Int? = null,
	val deletedRevision: String? = null,
)

@Serializable
data class MutationJournal(
	val id: String,
	val exchangeId: String,
	val messageId: String,
	val scope: String,
	val createdAt: Long,
	val records: List<MutationRecord>,
)

interface JournalStore {
	suspend fun saveJournal(journal: MutationJournal)

	suspend fun getJournal(id: String): MutationJournal?
}

data class RevertResult(
	val reverted: List<MutationRecord>,
	val conflicts: List<MutationRecord>,
	val remaining: List<MutationRecord>,
	val denied: Boolean = false,
)

data class AgentCapability(val writable: Boolean, val reason: String)

data class ToolCallRequest(val callId: String, val name: com.mdelacour.mynotes.ai.contract.ToolName, val arguments: kotlinx.serialization.json.JsonElement)

File: apps/web/src/lib/ai/journal.ts
export type MutationKind = 'edit' | 'create' | 'delete';
export type RevertState = 'pending' | 'reverted' | 'conflict' | 'dismissed';

export interface EditRecord {
	fromUtf16: number;
	toUtf16: number;
	deleted: string;
	inserted: string;
	startAnchor: unknown;
	endAnchor: unknown;
}

export interface MutationRecord {
	id: string;
	journalId: string;
	exchangeId: string;
	messageId: string;
	toolCallId: string;
	createdAt: number;
	kind: MutationKind;
	noteId: string;
	beforeRevision: string | null;
	afterRevision: string | null;
	revertState: RevertState;
	revertNote?: string;
	edits?: EditRecord[];
	created?: { content: string };
	deleted?: { content: string; orderIndex: number; revision: string };
}

export interface MutationJournal {
	id: string;
	exchangeId: string;
	messageId: string;
	scope: string;
	createdAt: number;
	records: MutationRecord[];
}

export interface RevertOutcome {
	reverted: MutationRecord[];
	conflicts: MutationRecord[];
	remaining: MutationRecord[];
}

export interface JournalStore {
	saveJournal(journal: MutationJournal): Promise<void>;
	getJournal(id: string): Promise<MutationJournal | undefined>;
	deleteJournalForMessage(messageId: string): Promise<void>;
}

export class InMemoryJournalStore implements JournalStore {
	private journals = new Map<string, MutationJournal>();

	async saveJournal(journal: MutationJournal): Promise<void> {
		this.journals.set(journal.id, structuredClone(journal));
	}

	async getJournal(id: string): Promise<MutationJournal | undefined> {
		const journal = this.journals.get(id);
		return journal ? structuredClone(journal) : undefined;
	}

	async deleteJournalForMessage(messageId: string): Promise<void> {
		for (const [id, journal] of this.journals) {
			if (journal.messageId === messageId) this.journals.delete(id);
		}
	}

	async list(): Promise<MutationJournal[]> {
		return [...this.journals.values()].map((journal) => structuredClone(journal));
	}
}
